src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs

```diff
@@ -55,57 +61,196 @@ internal static partial class JsonWriterHelper
+        private static readonly Vector128<sbyte> s_mask_SByte_0x26 = Vector128.Create((sbyte)0x26); // Ampersand '&'
+        private static readonly Vector128<sbyte> s_mask_SByte_0x27 = Vector128.Create((sbyte)0x27); // Apostrophe '''
+        private static readonly Vector128<sbyte> s_mask_SByte_0x2B = Vector128.Create((sbyte)0x2B); // Plus sign '+'
+        private static readonly Vector128<sbyte> s_mask_SByte_0x3C = Vector128.Create((sbyte)0x3C); // Less Than Sign '<'
+        private static readonly Vector128<sbyte> s_mask_SByte_0x3E = Vector128.Create((sbyte)0x3E); // Greater Than Sign '>'
+        private static readonly Vector128<sbyte> s_mask_SByte_0x5C = Vector128.Create((sbyte)0x5C); // Reverse Solidus '\'
+        private static readonly Vector128<sbyte> s_mask_SByte_0x60 = Vector128.Create((sbyte)0x60); // Grave Access '`'
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector128<short> CreateEscapingMask(Vector128<short> sourceValue)
         {
-            int idx;
+            Debug.Assert(Sse2.IsSupported);
 
-            if (encoder != null)
-            {
-                idx = encoder.FindFirstCharacterToEncodeUtf8(value);
-                goto Return;
-            }
+            Vector128<short> mask = Sse2.CompareLessThan(sourceValue, s_mask_UInt16_0x20); // Space ' ', anything in the control characters range
+
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x22)); // Quotation Mark '"'
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x26)); // Ampersand '&'
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x27)); // Apostrophe '''
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x2B)); // Plus sign '+'
+
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x3C)); // Less Than Sign '<'
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x3E)); // Greater Than Sign '>'
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x5C)); // Reverse Solidus '\'
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_UInt16_0x60)); // Grave Access '`'
+
+            mask = Sse2.Or(mask, Sse2.CompareGreaterThan(sourceValue, s_mask_UInt16_0x7E)); // Tilde '~', anything above the ASCII range
+
+            return mask;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector128<sbyte> CreateEscapingMask(Vector128<sbyte> sourceValue)
+        {
+            Debug.Assert(Sse2.IsSupported);
 
-            for (idx = 0; idx < value.Length; idx++)
+            Vector128<sbyte> mask = Sse2.CompareLessThan(sourceValue, s_mask_SByte_0x20); // Control characters, and anything above 0x7E since sbyte.MaxValue is 0x7E
+
+            mask = Sse2.Or(mask, Sse2.CompareEqual(sourceValue, s_mask_SByte_0x22)); // Quotation Mark "
```

Comment:
There are quite a lot comparisons and ors.
I have an approach that uses bit-masks, need to verify performance (later today).
It needs `Ssse3` though. But the same approach could be used for the squential part.

Maybe this makes more sense together with https://github.com/dotnet/corefx/pull/41845/files#r336728111

I'll give an update here.

Reply:
I implemented the idea for the byte-path (as it's easier, and to get some numbers quickly).

Code (base on current state of this PR): https://github.com/gfoidl/corefx/commit/5b5f8c6a8ecb95dca9559c40e3f1cdc483ed45e2

### Description

The basic idea is to to construct a bitmask, where `1` denotes the need for escaping, while `0` means no escaping needed. To select the correct bitmask, the input byte is split into the low- and high-nibble. Based on the low-nibble the bitmask is selected, and with the high-nibble the index in the bitmask is chosen.
One can illustrate this bitmask with the following image:
![grafik](https://user-images.githubusercontent.com/5755834/67148438-eb205f80-f29f-11e9-8265-d0f4892a91aa.png)

<details>
  <summary>code to genate the matrix shown</summary>

```c#
using System;
using System.Collections.Generic;

namespace ConsoleApp3
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("\t\t\t\t\t\t\t\t\thigh nibble");
            Console.Write("low nibble\t");
            for (int j = 0; j <= 0xF; ++j)
            {
                Console.Write($"{j:X}\t");
            }
            Console.WriteLine();
            for (int i = 0; i <= 0xF; ++i)
            {
                Console.Write($"\t{i:X}\t");

                for (int j = 0; j <= 0xF; ++j)
                {
                    int codePoint = j << 4 | i;

                    bool needsEscaping = NeedsEscaping(codePoint);
                    Console.ForegroundColor = needsEscaping
                        ? ConsoleColor.Red
                        : ConsoleColor.Green;

                    Console.Write($"{(needsEscaping ? "1" : "0")}\t");
                    Console.ResetColor();
                }

                Console.WriteLine();
            }
        }

        private static readonly HashSet<int> s_exclude = new HashSet<int>(new int[]
        {
            '"', '&', '\'', '+', '<', '>', '\\', '`'
        });

        private static bool NeedsEscaping(int value)
        {
            return value < ' '
                || value > '~'
                || s_exclude.Contains(value);
        }
    }
}
```
</details>

Based on this "matrix" the bitmask is created -> https://github.com/gfoidl/corefx/blob/5b5f8c6a8ecb95dca9559c40e3f1cdc483ed45e2/src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs#L156-L173

A sequential version of the code looks like:
```c#
private static bool NeedsEscaping(byte value)
{
    int highNibble = value >> 4;
    int lowNibble = value & 0xF;
    byte actualMask = Bitmask[lowNibble];

    bool needsEscaping = ((1 << highNibble) & actualMask) != 0 || highNibble > 0x7;

    return needsEscaping;
}
```

### Code-gen (differences)

The main-check-loop gets a lot smaller (I know, size is not the only measure):
```asm
G_M29314_IG07:      ;; bbWeight=4
       C4C17A6F2424         vmovdqu  xmm4, xmmword ptr [r12]
       C5D172D404           vpsrld   xmm5, xmm4, 4
       C5D1DBE8             vpand    xmm5, xmm5, xmm0
       C5D9DBE0             vpand    xmm4, xmm4, xmm0
       C4E27100E4           vpshufb  xmm4, xmm1, xmm4
       C4E26900ED           vpshufb  xmm5, xmm2, xmm5
       C5D1DBE4             vpand    xmm4, xmm5, xmm4
       C5D964E3             vpcmpgtb xmm4, xmm4, xmm3
       C5F9D7C4             vpmovmskb eax, xmm4
       85C0                 test     eax, eax
       7542                 jne      SHORT G_M29314_IG12

G_M29314_IG08:      ;; bbWeight=4
       4183C710             add      r15d, 16
       4983C410             add      r12, 16
       453BEF               cmp      r13d, r15d
       7DC6                 jge      SHORT G_M29314_IG07
```
as opposed to the masking-approach in this PR
```asm
G_M29314_IG07:      ;; bbWeight=4
       F3450F6F0C24         movdqu   xmm9, xmmword ptr [r12]
       440F28D0             movaps   xmm10, xmm0
       66450F64D1           pcmpgtb  xmm10, xmm9
       450F28D9             movaps   xmm11, xmm9
       66440F74D9           pcmpeqb  xmm11, xmm1
       66450FEBD3           por      xmm10, xmm11
       450F28D9             movaps   xmm11, xmm9
       66440F74DA           pcmpeqb  xmm11, xmm2
       66450FEBD3           por      xmm10, xmm11
       450F28D9             movaps   xmm11, xmm9
       66440F74DB           pcmpeqb  xmm11, xmm3
       66450FEBD3           por      xmm10, xmm11
       450F28D9             movaps   xmm11, xmm9
       66440F74DC           pcmpeqb  xmm11, xmm4
       66450FEBD3           por      xmm10, xmm11
       450F28D9             movaps   xmm11, xmm9
       66440F74DD           pcmpeqb  xmm11, xmm5
       66450FEBD3           por      xmm10, xmm11
       450F28D9             movaps   xmm11, xmm9
       66440F74DE           pcmpeqb  xmm11, xmm6
       66450FEBD3           por      xmm10, xmm11
       450F28D9             movaps   xmm11, xmm9
       66440F74DF           pcmpeqb  xmm11, xmm7
       66450FEBD3           por      xmm10, xmm11
       66450F74C8           pcmpeqb  xmm9, xmm8
       66450FEBD1           por      xmm10, xmm9
       66410FD7C2           pmovmskb  eax, xmm10
       85C0                 test     eax, eax
       7540                 jne      SHORT G_M29314_IG12

G_M29314_IG08:      ;; bbWeight=4
       4183C710             add      r15d, 16
       4983C410             add      r12, 16
       453BEF               cmp      r13d, r15d
       0F8D6BFFFFFF         jge      G_M29314_IG07
```

### Benchmark-Results

Benchmark-code based on https://gist.github.com/ahsonkhan/c566f5e7d65c1fde5a83a67be290c4ee#file-test_escapingbenchmark-cs

In the results I stripped of the ones for the sequential path, as they are equal (it's the same code-path used).

```
|                 Method | TestStringData |      Mean |     Error |    StdDev | Ratio | RatioSD |
|----------------------- |--------------- |----------:|----------:|----------:|------:|--------:|
| NeedsEscapingCurrentPR |        100, -1 | 36.110 ns | 0.3919 ns | 0.3474 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |        100, -1 | 26.421 ns | 0.4391 ns | 0.4107 ns |  0.73 |    0.02 |
|                        |                |           |           |           |       |         |
| NeedsEscapingCurrentPR |         15, -1 | 21.821 ns | 0.1093 ns | 0.1022 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |         15, -1 | 21.769 ns | 0.0974 ns | 0.0863 ns |  1.00 |    0.01 |
|                        |                |           |           |           |       |         |
| NeedsEscapingCurrentPR |         16, -1 | 13.148 ns | 0.0447 ns | 0.0396 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |         16, -1 | 11.580 ns | 0.0190 ns | 0.0168 ns |  0.88 |    0.00 |
|                        |                |           |           |           |       |         |
| NeedsEscapingCurrentPR |         17, -1 | 13.861 ns | 0.0410 ns | 0.0342 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |         17, -1 | 12.344 ns | 0.0284 ns | 0.0237 ns |  0.89 |    0.00 |
|                        |                |           |           |           |       |         |
| NeedsEscapingCurrentPR |         32, -1 | 17.290 ns | 0.0554 ns | 0.0491 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |         32, -1 | 13.878 ns | 0.0854 ns | 0.0799 ns |  0.80 |    0.01 |
|                        |                |           |           |           |       |         |
| NeedsEscapingCurrentPR |          9, -1 | 16.961 ns | 0.0853 ns | 0.0798 ns |  1.00 |    0.00 |
|   NeedsEscapingBitMask |          9, -1 | 16.855 ns | 0.0837 ns | 0.0742 ns |  0.99 |    0.01 |
```

@ahsonkhan how did you run the benchmarks with escaping needed? (I'm stuck in the benchmark-code and don't want to explode the things :wink:)

### Discussion

The results meet my expectations so far, as less code is involved.
It add a bit of complexity, as there is a path for SSE2 and one for SSSE3.

Is this something to pursue?

If this should be productized, more comments are needed, of course.

Reply:
> how did you run the benchmarks with escaping needed?

I ran this benchmark.
https://gist.github.com/ahsonkhan/c566f5e7d65c1fde5a83a67be290c4ee#file-test_escapingbenchmark-cs-L1

For input lengths 1, 7, and 8, it adds a character that needs to be escaped at index i (from 0 to length).
https://gist.github.com/ahsonkhan/c566f5e7d65c1fde5a83a67be290c4ee#file-test_escapingbenchmark-cs-L31-L32